core/src/test/java/kafka/test/ClusterInstance.java

```diff
@@ -0,0 +1,98 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.test;
+
+import kafka.network.SocketServer;
+import kafka.test.annotation.ClusterTest;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.common.network.ListenerName;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Properties;
+
+public interface ClusterInstance {
+
+    enum ClusterType {
+        Zk,
+        // Raft
+    }
+
+    /**
+     * Cluster type. For now, only ZK is supported.
+     */
+    ClusterType clusterType();
+
+    /**
+     * The cluster configuration used to create this cluster. Changing data in this instance through this accessor will
+     * have no affect on the cluster since it is already provisioned.
+     */
+    ClusterConfig config();
+
+    /**
+     * The listener for this cluster as configured by {@link ClusterTest} or by {@link ClusterConfig}. If
+     * unspecified by those sources, this will return the listener for the default security protocol PLAINTEXT
+     */
+    ListenerName listener();
+
+    /**
+     * The broker connect string which can be used by clients for bootstrapping
+     */
+    String brokerList();
+
+    /**
+     * A collection of all brokers in the cluster. In ZK-based clusters this will also include the broker which is
+     * acting as the controller (since ZK controllers serve both broker and controller roles).
+     */
+    Collection<SocketServer> brokers();
+
+    /**
+     * A collection of all controllers in the cluster. For ZK-based clusters, this will return the broker which is also
+     * currently the active controller. For Raft-based clusters, this will return all controller servers.
+     */
+    Collection<SocketServer> controllers();
+
+    /**
+     * Any one of the broker servers.
+     */
+    Optional<SocketServer> anyBroker();
+
+    /**
+     * Any one of the controller servers.
+     */
+    Optional<SocketServer> anyController();
+
+    /**
+     * The underlying object which is responsible for setting up and tearing down the cluster.
+     */
+    Object getUnderlying();
+
+    default <T> T getUnderlying(Class<T> asClass) {
+        return asClass.cast(getUnderlying());
+    }
+
+    Admin createAdminClient(Properties configOverrides);
+
+    default Admin createAdminClient() {
+        return createAdminClient(new Properties());
+    }
+
+    void start();
+
+    void stop();
```

Comment:
should this implement `AutoCloseable`?  Does code need to remember to close this cluster?

Or is that handled automatically by the framework...

Reply:
Originally I left these out of the interface because the framework handle starting and stopping the cluster. However, after @ijuma's feedback, I added `boolean autoStart` to ClusterConfig along with these methods so that a test could control cluster startup/shutdown explicitly. 

If a test does not call `stop`, the framework will stop the cluster. Stopping it a second time does nothing (per the implementation).

Reply:
ok